eluthia/zipapp_script.py

```diff
@@ -0,0 +1,187 @@
+#!/usr/bin/env python3
+from apps import config
+import subprocess
+import zipfile
+import os
+import time
+
+
+class InstallationError(Exception): # Used to signify that installation / verification failed and we should roll back
+    pass
+
+
+def get_version(package, deb_file=False):
+    """
+    Get the version of a package or verify that it is installed
+    Input:
+        string package - Either the path to a debian package file, or the name of an installed package
+        bool deb_file - set to True if inputting the path to a debian package file
+    Output:
+        string version - version of the package
+            NoneType None denotes that the package is not installed.
+    """
+    if deb_file:
+        return subprocess.check_output(['dpkg-deb', '-f', package, 'Version']).decode().rstrip()
+
+    try:
+        return subprocess.check_output(['dpkg-query', '--showformat=\'${Version}\'', '--show', package]).decode().strip('\'')
+    except subprocess.CalledProcessError: # This happens if the package is not installed.
+        return None
+
+
+def unpack_zip(zip_path, unpack_path):
+    """
+    Unpacks the packages in the archive into NEW_DEB_PATH. Also extracts information about the packages.
+    Input:
+        string zip_path - path to the archive to be extracted (this should always be os.path.dirname(__file__), but it's a parameter for modularity)
+        string unpack_path - path to where the packages should be placed
+    Output:
+        list package_list - list of packages inside the archive.
+    """
+    with zipfile.ZipFile(zip_path) as z:
+        package_list = [item for item in z.namelist() if item.lower().endswith('.deb')]
+
+        for package in package_list:
+            z.extract(package, unpack_path)
+
+    return package_list
+
+
+def repack(package_dict, repack_path):
+    """
+    Repackage installed debian packages into .deb files
+    Input:
+        dict package_dict - a dictionary of packages to repack - Format: {package_name: version}
+        string repack_path - path to where repackaged .deb files should be placed
+    """
+    if get_version('dpkg_repack') is None:
+        subprocess.run(['apt-get', 'install', 'dpkg-repack'])
+
+    for package_name, version in package_dict.items():
+        subprocess.run(['dpkg-repack', package_name], cwd=repack_path)
+        os.rename(f'{repack_path}/{package_name}_{version}_all.deb', f'{package_name}.deb')
```

Comment:
I don't know if that's a safe assumption that that's the output DEB file name it will always end up with. Isn't better to `os.listdir` the content of `repack_path`, and that will give a list of paths to the DEB files?

And then `repack` will just need `package_dict` to just be `package_names`.